Modularity: Metrics infrastructure and collection

legendko opened the pull request.

SyntaxNode commented on metrics/metrics.go:
```diff
@@ -49,6 +49,12 @@ type PrivacyLabels struct {
+type ModuleLabels struct {
+	Module string
+	Stage  string
+	PubID  string // account ID
```
Please use standard Go comment style, in this case something like "// PubID is the account id", or just rename `PubID` to `AccountID` if you'd like.

legendko replied:
renamed to `AccountID`

SyntaxNode commented on hooks/hookexecution/execution.go:
```diff
@@ -166,8 +170,13 @@ func handleHookResponse[P any](
 	var rejectErr *RejectError
+	labels := metrics.ModuleLabels{Module: hr.HookID.ModuleCode, Stage: ctx.stage, PubID: ctx.accountId}
+	metricEngine.RecordModuleCalled(labels)
+	metricEngine.RecordModuleDuration(labels, hr.ExecutionTime)
```
I recommend combining `RecordModuleCalled` and `RecordModuleDuration` since they're only called here once as a pair and it makes sense IMHO for a call to be directly associated with a duration.

legendko replied:
makes sense, updated

SyntaxNode commented on metrics/go_metrics.go:
```diff
@@ -215,13 +238,25 @@ func NewBlankMetrics(registry metrics.Registry, exchanges []openrtb_ext.BidderNa
+	}
+
+	return mn
+}
```
I recommend changing the `mn` variable to `names` for improved readability.

AlexBVolcy replied:
Yeah I agree with the need to change the variable names here

SyntaxNode commented on metrics/go_metrics.go:
```diff
@@ -781,3 +873,142 @@ func (me *Metrics) RecordAdsCertReq(success bool) {
+
+	// Module metrics
+	mm.DurationTimer.Update(duration)
+	// Account-Module metrics
```
Please add a line break before the Account-Module comment throughout to improve readability.

SyntaxNode commented on metrics/prometheus/preload.go:
```diff
@@ -207,6 +208,40 @@ func preloadLabelValues(m *Metrics, syncerKeys []string) {
+
+	for module := range moduleStageNames {
+		preloadLabelValuesForHistogram(m.moduleDuration[module], map[string][]string{
+			stageLabel: stageValues,
```
Shouldn't this look more like:

```
for module, stageValues := range moduleStageNames {
  preloadLabelValuesForHistogram(m.moduleDuration[module], map[string][]string{
    stageLabel: stageValues,
```

This way we only pre-create labels which are needed for the particular module.

.. or else just use all possible stage values instead.

legendko replied:
I agree with you. I made that all possible stages are added for every module, but that's unnecessary. Fixed

SyntaxNode commented on metrics/prometheus/type_conversion.go:
```diff
@@ -130,3 +130,20 @@ func tcfVersionsAsString() []string {
+	}
+
+	return stages
+}
```
Please see my earlier comment. This method is likely no longer necessary.

SyntaxNode commented on modules/helpers.go:
```diff
@@ -0,0 +1,72 @@
+}
+
+func addModuleStageName(moduleStageNameCollector map[string][]string, id string, stage string) map[string][]string {
+	replacer := strings.NewReplacer(".", "_", "-", "_")
```
Can you move the replacer to a top level package scoped variable to save on some allocations?

legendko replied:
Done

AlexBVolcy commented on endpoints/openrtb2/amp_auction.go:
```diff
@@ -77,7 +77,7 @@ func NewAmpEndpoint(
 	}
 
-	hookExecutor := hookexecution.NewHookExecutor(hookExecutionPlanBuilder, hookexecution.EndpointAmp)
+	hookExecutor := hookexecution.NewHookExecutor(hookExecutionPlanBuilder, hookexecution.EndpointAmp, met)
```
Would it be possible to update this name to something more descriptive like `metrics` or `metricsEngine`?

AlexBVolcy commented on endpoints/openrtb2/auction.go:
```diff
@@ -86,7 +86,7 @@ func NewEndpoint(
 	}
 
-	hookExecutor := hookexecution.NewHookExecutor(hookExecutionPlanBuilder, hookexecution.EndpointAuction)
+	hookExecutor := hookexecution.NewHookExecutor(hookExecutionPlanBuilder, hookexecution.EndpointAuction, met)
```
Would it be possible to update this name to something more descriptive like `metrics` or `metricsEngine`?

AlexBVolcy commented on metrics/go_metrics.go:
```diff
@@ -311,6 +346,10 @@ func NewMetrics(registry metrics.Registry, exchanges []openrtb_ext.BidderName, d
 	newMetrics.AdsCertRequestsFailure = metrics.GetOrRegisterMeter("ads_cert_requests.failed", registry)
 	newMetrics.adsCertSignTimer = metrics.GetOrRegisterTimer("ads_cert_sign_time", registry)
 
+	for m, stages := range moduleStageNames {
```
Could you rename `m` to be `modules`

AlexBVolcy commented on metrics/go_metrics.go:
```diff
@@ -342,6 +381,28 @@ func makeBlankAdapterMetrics(disabledMetrics config.DisabledMetrics) *AdapterMet
 }
 
+func makeBlankModuleStageMetrics(stages []string) map[string]*ModuleMetrics {
+	mm := map[string]*ModuleMetrics{}
```
Could this be renamed for readability?

AlexBVolcy commented on metrics/go_metrics.go:
```diff
@@ -781,3 +873,142 @@ func (me *Metrics) RecordAdsCertReq(success bool) {
+
+	// Module metrics
+	mm.CallCounter.Inc(1)
+	// Account-Module metrics
```
Agree there should be a line break between line 900 and 901

AlexBVolcy commented on metrics/go_metrics.go:
```diff
@@ -781,3 +873,142 @@ func (me *Metrics) RecordAdsCertReq(success bool) {
+
+	// Module metrics
+	mm.FailureCounter.Inc(1)
+	// Account-Module metrics
```
Same line break needed

AlexBVolcy commented on metrics/go_metrics.go:
```diff
@@ -781,3 +873,142 @@ func (me *Metrics) RecordAdsCertReq(success bool) {
+
+	// Module metrics
+	mm.SuccessNoopCounter.Inc(1)
+	// Account-Module metrics
```
Same line break needed

AlexBVolcy reviewed (commented):
Minor comments related to variable re-naming and small changes

legendko commented:
> Minor comments related to variable re-naming and small changes

Updated

AlexBVolcy commented on modules/modules.go:
```diff
@@ -50,18 +52,25 @@ func (m *builder) Build(cfg config.Modules, client *http.Client) (hooks.HookRepo
 	}
 
-	return hooks.NewHookRepository(modules)
+	coll, err := createModuleStageNamesCollection(modules)
```
Nitpick: Could you re-name `coll` to be more descriptive

legendko replied:
done

AlexBVolcy reviewed (commented):
Looks good, almost ready to approve, just left one small comment

SyntaxNode commented on metrics/go_metrics.go:
```diff
@@ -383,6 +444,30 @@ func registerAdapterMetrics(registry metrics.Registry, adapterOrAccount string,
+	mm.FailureCounter = metrics.GetOrRegisterCounter(fmt.Sprintf("account.%s.modules.module.%s.failure", id, module), registry)
+	mm.SuccessNoopCounter = metrics.GetOrRegisterCounter(fmt.Sprintf("account.%s.modules.module.%s.success.noop", id, module), registry)
+	mm.SuccessUpdateCounter = metrics.GetOrRegisterCounter(fmt.Sprintf("account.%s.modules.module.%s.success.update", id, module), registry)
+	mm.SuccessRejectCounter = metrics.GetOrRegisterCounter(fmt.Sprintf("account.%s.modules.module.%s.reject", id, module), registry)
```
Good catch. It should be `success.reject`. I can see how `success.reject` looks strange, but it makes sense in the module ran successfully and then decided to reject the request.

legendko replied:
eagle eye :)

legendko commented:
Latest updates:
1) Removed metrics gathering for `unknown` accounts
2) Added condition to determine whether the module was really successful
3) Fixes for your comments

bsardo commented on hooks/hookexecution/execution.go:
```diff
@@ -264,6 +296,16 @@ func handleHookMutations[P any](payload P, hr hookResponse[P], hookOutcome *Hook
+		successfulMutations++
+	}
+
+	// if at least one mutation from a given module was successfully applied
+	// we consider that the module was processed successfully
+	if successfulMutations > 0 {
+		metricEngine.RecordModuleSuccessUpdated(labels)
+	} else {
+		hookOutcome.Status = StatusExecutionFailure
+		metricEngine.RecordModuleExecutionError(labels)
 	}
```
I don't see anything in the PRD about making this determination. Is this also PBS-Java's behavior?

bsardo replied:
I guess this makes sense since this logic will only execute if there's at least one mutation and if at least one mutation is successfully applied we consider that an update. If there were also failed updates, they would be wrapped up in the hook outcome `Warnings`.

legendko replied:
No, this is our own assumption based on our previous discussions on the topic what should we do if a module can not apply a mutation because another module deleted some property. We discussed that the client is responsible for a correct module configuration and should check by himself that the modules do not operate with the same fields. In case such event occurs we discussed that we should simply log a warning. But we realized that we could have a rare occasion when a module has only 1 mutation and it can not be applied (or N mutations and all of them can not be applied), then it seems strange that we send a successful metric and record a successful status for that module. If you think we should revert the changes or it should have another behavior please correct us.

AlexBVolcy reviewed (approved):
LGTM
